File: packages/sdk/src/core/adapters/storage-adapter.ts
import type { Result } from '../types';

export type StorageAdapter = {
  getItem<T>(key: string): Promise<Result<T | null>>;
  setItem<T>(key: string, value: T): Promise<Result<void>>;
  removeItem(key: string): Promise<Result<void>>;
  clear(): Promise<Result<void>>;
};

File: packages/swift/Sources/PhaseAnalytics/Core/Adapters/StorageAdapter.swift
import Foundation

public protocol StorageAdapter: Sendable {
    func getItem<T: Codable & Sendable>(key: String) async -> Result<T?, Error>
    func setItem<T: Codable & Sendable>(key: String, value: T) async -> Result<Void, Error>
    func removeItem(key: String) async -> Result<Void, Error>
    func clear() async -> Result<Void, Error>
}
